flow/src/main/java/com/sequenceiq/flow/service/FlowService.java

```diff
@@ -130,15 +133,24 @@ public FlowCheckResponse getFlowChainState(String chainId) {
             LOGGER.info("Checking if there is an active flow based on flow chain id {}", chainId);
             List<FlowChainLog> relatedChains = flowChainLogService.getRelatedFlowChainLogs(flowChains);
             Set<String> relatedChainIds = relatedChains.stream().map(FlowChainLog::getFlowChainId).collect(toSet());
-            Set<String> relatedFlowIds = flowLogDBService.getFlowIdsByChainIds(relatedChainIds);
-            List<FlowLogWithoutPayload> relatedFlowLogs = flowLogDBService.getFlowLogsWithoutPayloadByFlowIdsCreatedDesc(relatedFlowIds);
+            List<FlowLogWithoutPayload> relatedFlowLogs = flowLogDBService.getFlowLogsWithoutPayloadByFlowChainIdsCreatedDesc(relatedChainIds);
             flowCheckResponse.setHasActiveFlow(!completed("Flow chain", chainId, relatedChains, relatedFlowLogs));
             flowCheckResponse.setLatestFlowFinalizedAndFailed(isFlowInFailedState(relatedFlowLogs, failHandledEvents));
-            return flowCheckResponse;
+            if (!flowCheckResponse.getHasActiveFlow() && !relatedFlowLogs.isEmpty()) {
+                List<FlowLogWithoutPayload> flowLogForEndTime = relatedFlowLogs.stream().filter(s -> !Objects.equals(s.getCurrentState(), "FINISHED"))
```

Comment:
Could you please split the below logic into smaller logic. As I see the main point is to get the end time. Other things are just implementation detail.